File: app/src/main/java/com/semye/android/xml/DomXml.java
package com.semye.android.xml;


import androidx.annotation.NonNull;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;


public class DomXml {
    private Document doc;

    public DomXml(InputStream is) {
        try {
            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            DocumentBuilder builder = factory.newDocumentBuilder();
            doc = builder.parse(is);
        } catch (@NonNull ParserConfigurationException | SAXException | IOException e) {
            e.printStackTrace();
        }
    }

    @NonNull
    public List<HashMap<String, String>> getProducts() {
        Element root = doc.getDocumentElement();
        NodeList lstNodes = root.getElementsByTagName("product");
        List<HashMap<String, String>> products = new ArrayList<>();
        for (int i = 0; i < lstNodes.getLength(); i++) {
            HashMap<String, String> map = new HashMap<>();
            Element element = (Element) lstNodes.item(i);
            Element elemId = (Element) element.getElementsByTagName("id").item(0);
            map.put("id", elemId.getTextContent());
            Element elemName = (Element) element.getElementsByTagName("name").item(0);
            map.put("name", elemName.getTextContent());
            Element elemPrice = (Element) element.getElementsByTagName("price").item(0);
            map.put("price", elemPrice.getTextContent());
            products.add(map);
        }
        return products;
    }

}

File: app/src/main/java/com/semye/android/ui/xml/DomXml.kt
package com.semye.android.ui.xml

import org.w3c.dom.Document
import org.w3c.dom.Element
import org.xml.sax.SAXException
import java.io.IOException
import java.io.InputStream
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

class DomXml(inputStream: InputStream?) {
    private var doc: Document? = null
    val products: List<HashMap<String, String>>
        get() {
            val root = doc!!.documentElement
            val lstNodes = root.getElementsByTagName("product")
            val products: MutableList<HashMap<String, String>> = ArrayList()
            for (i in 0 until lstNodes.length) {
                val map = HashMap<String, String>()
                val element = lstNodes.item(i) as Element
                val elemId = element.getElementsByTagName("id").item(0) as Element
                map["id"] = elemId.textContent
                val elemName = element.getElementsByTagName("name").item(0) as Element
                map["name"] = elemName.textContent
                val elemPrice = element.getElementsByTagName("price").item(0) as Element
                map["price"] = elemPrice.textContent
                products.add(map)
            }
            return products
        }

    init {
        try {
            val factory = DocumentBuilderFactory.newInstance()
            val builder = factory.newDocumentBuilder()
            doc = builder.parse(inputStream)
        } catch (e: ParserConfigurationException) {
            e.printStackTrace()
        } catch (e: SAXException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}